### DISCUSSION TITLE
Prisma + Supabase: Create RLS Policies?

### DISCUSSION THREAD

**Discussion by AdamEisfeld** [prisma/prisma#18642] [Q&A] [answered]
I'm a frontend developer that has been spending the last few months trying to brush up on my backend chops, please bare with me.

I am currently trying to figure out a stack for my project, and am leaning towards (what I gather is not a common use):
- using Supabase for the db
- using Supabase edge functions as an "api" I expose to my frontend (one edge function, which reads in params from the request to route it to the appropriate handler)
- using Supabase-js client inside edge functions for reading to / writing from the db (I'd prefer to use Prisma client but I haven't found much info on getting Prisma client to play nicely inside a Supabase edge function)
- using Prisma migrations for defining my Supabase db structure (I don't like the idea of using a visual editor on Supabase for this, I would rather use Prisma and have the single source of truth live in my repo where devs can see what is going on quickly).
- using Zod for type safety when working with Supabase-js client output (again, I'd rather use Prisma's type definitions / client but, this works for now)

I've got an MVP up and running and it seems to work, but I want to ensure at least 1 RLS policy is enforced for every table I create, requiring the user to be authenticated before allowing them to read / write / delete / update.

I am able to do this by manually modifying a migration file created from Prisma before migrating the changes to Supabase, eg I can go in and add:

```
-- AddRLS teams
ALTER TABLE "teams" ENABLE ROW LEVEL SECURITY;
CREATE POLICY "Require auth" ON "teams" AS PERMISSIVE FOR INSERT TO authenticated WITH CHECK (true);
```

But this is error prone and tedious. Is there a way to declare these RLS policies on my tables inside my schema.prisma file?
- Adam

**Comment by AdamEisfeld** [comment, marked as answer]
For the time being I've created a little helper script for adding these RLS / Policies as well as any additional idempotent SQL statements to a new migration file. Designed to be part of the CI pipeline, just after the latest db migration file has been generated, but just before deploying those migrations to the db.

The script first searches through all of the prisma migration files, building up a list of every table name in the db by searching for all "CREATE TABLE" statements (I need to add functionality to track DROP TABLE / RENAME TABLE still).

The script accepts a single parameter specifying the path to a config file. The config file should specify an array of "tableAlterations" A tableAlteration specifies a regex string that should output a table name from a given SQL statement. A list is built up of each table name that is not found for each tableAlteration.

The tableAlteration should also specify an "alteration" value, which is what will be inserted into the SQL if it is missing. This is an array of strings. If a given entry in the array ends with ".sql" then the corresponding sql file will be used in place of it, otherwise the string itself will be used.

Finally, the config can also specify "additionalInsertions", which again is an array of strings, where each string can either be the path to an .sql file or a blob of SQL to insert. These additionalInsertions should be idempotent and are designed to be inserted on every DB deploy. I plan on using this for ensuring my Supabase Database Functions are always pushed.

Here is an example config file which ensures every table in the DB has RLS enabled and has an RLS policy that requires authentication for all operations. It also ensures a Database Function is created that causes User signups to result in a new row inserted into a "profiles" table with the user's id / first / last name / email.

```json
{
	"tableAlterations": [
		{
			"regex": "ALTER TABLE\\s+[\"]?([^\\s(\"]+)[\"]?\\s+ENABLE\\s+ROW\\s+LEVEL\\s+SECURITY;",
			"alterations": [
				"-- Enable RLS for \"$1\"",
				"ALTER TABLE \"$1\" ENABLE ROW LEVEL SECURITY;"
			]
		},
		{
			"regex": "CREATE POLICY\\s+\".*?\"\\s+ON\\s+\"(.*?)\"",
			"alterations": [
				"-- Create RLS Policy for \"$1\"",
				"CREATE POLICY \"Must be authenticated\" ON \"$1\"",
				"AS PERMISSIVE FOR ALL",
				"TO authenticated",
				"WITH CHECK (true);"
			]
		}
	],
	"additionalInsertions": [
		"prisma/handle_new_user.sql"
	]
}
```

Here is the handle_new_user.sql file:

```json
-- inserts a row into public.profiles
create or replace function public.handle_new_user()
returns trigger
language plpgsql
security definer set search_path = public
as $$
begin
	insert into public.profiles (uuid, email)
	values (new.id, new.email);
	return new;
end;
$$;

-- trigger the function every time a user is created
create trigger on_user_created
	after insert on auth.users
	for each row execute procedure public.handle_new_user();
```

After running on my starter project that contains tables for "teams", "profiles", and "team_profiles", I get this additional migration created in the prisma/migrations folder:

```sql
-- Generated for config prisma/migrate_config.json
-- Thu Apr 06 2023 01:57:08 GMT-0400 (Eastern Daylight Time)

-- Enable RLS for "profiles"
ALTER TABLE "profiles" ENABLE ROW LEVEL SECURITY;

-- Enable RLS for "team_profiles"
ALTER TABLE "team_profiles" ENABLE ROW LEVEL SECURITY;

-- Enable RLS for "teams"
ALTER TABLE "teams" ENABLE ROW LEVEL SECURITY;

-- Create RLS Policy for "profiles"
CREATE POLICY "Must be authenticated" ON "profiles"
AS PERMISSIVE FOR ALL
TO authenticated
WITH CHECK (true);

-- Create RLS Policy for "team_profiles"
CREATE POLICY "Must be authenticated" ON "team_profiles"
AS PERMISSIVE FOR ALL
TO authenticated
WITH CHECK (true);

-- Create RLS Policy for "teams"
CREATE POLICY "Must be authenticated" ON "teams"
AS PERMISSIVE FOR ALL
TO authenticated
WITH CHECK (true);

-- inserts a row into public.profiles
create or replace function public.handle_new_user()
returns trigger
language plpgsql
security definer set search_path = public
as $$
begin
	insert into public.profiles (uuid, email)
	values (new.id, new.email);
	return new;
end;
$$;

-- trigger the function every time a user is created
create trigger on_user_created
	after insert on auth.users
	for each row execute procedure public.handle_new_user();

```


And here is the script in question (save as .js or .cjs, run via node). I have it set up as an npm script to run via ```npm run post-migrate prisma/post_migrate_config.json```:

```js
#! /usr/bin/env node

// -------------------
// Helpers
// -------------------

const run = (command, inherit = true) => {
	const child_process = require('child_process')
	if (inherit) {
		return child_process.execSync(command, { stdio: 'inherit' })
	} else {
		return child_process.execSync(command).toString()
	}
}

const writeFile = (data, path) => {
	const createDirectories = (pathname) => {
		const path = require('path')
		const fs = require('fs')
		const __dirname = path.resolve()
		pathname = pathname.replace(/^\.*\/|\/?[^\/]+\.[a-z]+|\/$/g, '') // Remove leading directory markers, and remove ending /file-name.extension
		if (pathname.length > 0) {
			fs.mkdirSync(pathname, { recursive: true })
		}
	}
	createDirectories(path)
	const fs = require('fs')
	
	fs.writeFileSync(path, data)
}

const readFile = (path) => {
	const fs = require('fs')
	return fs.readFileSync(path)
}

// -------------------
// Classes
// -------------------

class MigrationFile {

	constructor(path) {
		this.path = path
		this.sql = readFile(path, 'utf8')
		this.tablesCreated = this.getCreatedTableNames()
	}

	getCreatedTableNames() {
		return this.getRegexTableNames('CREATE\\s+TABLE\\s+["]?([^\\s("]+)["]?')
	}

	getRegexTableNames(regexString) {
		const regex = new RegExp(regexString, 'gi') // regular expression to match "ALTER TABLE" statements enabling RLS and capture the table name
		const tableNames = []
		let match
		while ((match = regex.exec(this.sql)) !== null) {
			tableNames.push(match[1]) // add the captured table name to the array
		}
		return tableNames
	}

}

// -------------------
// Script
// -------------------

// Setup
const args = process.argv?.slice(2)
const configPath = args[0]
const config = configPath ? JSON.parse(readFile(configPath)) : {}
const tableNames = []
const migrations = []
const generatedSQL = []

console.log('Running post-migration script...')

// Find all of our migration files
const migrationFilePaths = (run('ls prisma/migrations/*/**.sql', false) ?? '').trim().split('\n')

console.log(`Found migration files: ${migrationFilePaths.length}`)

// Loop through each migration file building up a list of tables
for (const migrationFilePath of migrationFilePaths) {

	// Read the SQL file
	const migrationFile = new MigrationFile(migrationFilePath)

	// Add to migrations
	migrations.push(migrationFile)

	// Build list of all table names without dupes
	for (const tableName of migrationFile.tablesCreated) {
		if (!tableNames.includes(tableName)) {
			tableNames.push(tableName)
		}
	}

}

// Loop over each table alteration, checking if it is missing for any of our tables across all files
for (const tableAlteration of config.tableAlterations ?? []) {

	// Check if the table alteration is missing for any of our tables
	const missingTables = tableNames.filter(tableName => !migrations.some(migration => migration.getRegexTableNames(tableAlteration.regex).includes(tableName)))

	// If it is missing for any of our tables, add it to the generated SQL file
	if (missingTables.length > 0) {

		for (const tableName of missingTables) {
			
			for (const alteration of tableAlteration.alterations) {

				if (alteration.endsWith('.sql')) {
					// If it is an SQL file, read it and add it to the generated SQL file
					generatedSQL.push(readFile(alteration, 'utf8').toString().replaceAll('$1', tableName))
				} else {
					// Otherwise, just add it to the generated SQL file
					generatedSQL.push(alteration.replaceAll('$1', tableName))
				}
				
			}

			generatedSQL.push('')

		}

	}

}

// Loop over each additional insertion, adding it to the generated SQL file
for (const additionalInsertion of config.additionalInsertions ?? []) {

	if (additionalInsertion.endsWith('.sql')) {
		// If it is an SQL file, read it and add it to the generated SQL file
		generatedSQL.push(readFile(additionalInsertion, 'utf8').toString())
	} else {
		// Otherwise, just add it to the generated SQL file
		generatedSQL.push(additionalInsertion)
	}

	generatedSQL.push('')

}

// Exit early if needed
if (generatedSQL.length === 0) {
	console.log('No additional SQL to add. ✅')
	return
}

console.log(`Found ${generatedSQL.length} additional lines of SQL to add. 📦`)

// Insert meta into SQL
generatedSQL.unshift(...[
	`-- Generated for config ${configPath}`,
	`-- ${new Date()}`,
	'',
])

// Write the generated SQL file
writeFile(generatedSQL.join('\n'), `prisma/migrations/${migrationFilePaths.length}_auto/migration.sql`)

```


Still somewhat rough for now but it seems to be doing it's job.

**Comment by ciekawy** [comment]
this is great but instead of closing should this be linked to some existing issue to make sure it can help with composing some complete solution?
https://github.com/prisma/prisma/issues/12735

also here is some related info
https://github.com/dthyresson/prisma-extension-supabase-rls/issues/2
